### ISSUE TITLE
Feature request: respect meta.version in fetchurl output hash

### ISSUE THREAD

**Issue by danbst** [NixOS/nixpkgs#43510]
## Issue description
Sometimes Nix users do a mistake - when upgrading package they update only version/url and forget to change hash (`sha256`). The effect is that package is not updated.

While changed `url` shouldn't trigger any warnings by design, situation with version isn't that clear.

## Solution
Add `version` more love.

1. Make it common to specify source version via `meta.version`. Example of derivation with new style:
```
{ stdenv, fetchurl
}:

let
    src = fetchurl rec {
        meta.version = "1.2.0";
        url = "https://github.com/graknlabs/grakn/releases/download/v${meta.version}/grakn-dist-${meta.version}.tar.gz";
        sha256 = "0hjsfj0m80nhd6mdj2m3razamqhqavpvjx80v1nsgfrypp4lr88f";
    };
in
stdenv.mkDerivation {
    name = "grakn-${src.meta.version}";
    inherit src;
    inherit (src.meta) version;
    meta.version = src.meta.version;
}
```
I think most of fetchurl-s in Nixpkgs may get version attribute.
2. When `meta.version` is specified, change `outputHashMode` to "recursive" and use `meta.version` in hash computation.
3. Now, package *source* will be responsible for specifying package version. And changing *source* version will require changes to *source* hash. 
4. Out of added benefit, it will now be unneeded to change 1 symbol of hash to trigger redownload/rehash :smiley: 
5. Think of something similar for `fetchFromGithub` and co.

**Comment by volth** [comment]
maybe
```
# imaginary library function (String -> String -> String)
strcrypt = key: data: blablabla;

stdenv.mkDerivation {
  version = "1.0"
  name = "grakn-${version}";
  src = fetchurl {
    url = "https://github.com/graknlabs/grakn/releases/download/v${version}/grakn-dist-${version}.tar.gz";
    sha256 = strcrypt version "0hjsfj0m80nhd6mdj2m3razamqhqavpvjx80v1nsgfrypp4lr88f";
  };
};
```

the only addition is ```strcrypt version ``` and it is flexible what is included in the key

**Comment by danbst** [comment]
@volth that `strcrypt` should be reversible

1. update version
2. build
3. observe hash mismatch
4. Convert hash from error message using `(strcrypt name)^(-1)` into input hash.

Or maybe I don't see something?

**Comment by volth** [comment]
yes, I meant symmetric crypto, so ```(strcrypt key (strcrypt key a)) == a```

**Comment by danbst** [comment]
That breaks the current build-fail-copy-paste hash update by adding extra hash convert step. However we can add new outputHashAlgo
```
 src = fetchurl {
    url = "https://github.com/graknlabs/grakn/releases/download/v${version}/grakn-dist-${version}.tar.gz";
    outputHashAlgo = "scrypt-sha256";
    scrypt-sha256 = strcrypt name "0hjsfj0m80nhd6mdj2m3razamqhqavpvjx80v1nsgfrypp4lr88f";
  };
```
But even then, fetchurl should know `name` argument.

On other note, it will promote usage of automated updaters.

**Comment by edolstra** [comment]
`fetchurl` is definitely still a fixed-output derivation. The third example triggers a rebuild because the `name` attribute (and therefore the store path) has changed.

**Comment by danbst** [comment]
Oh, cool! But I'll leave issue closed for now, don't want to push more here.

**Comment by danbst** [comment]
upd on this small RFE. I've merged https://github.com/NixOS/nixpkgs/pull/66173, which made `meta.version` from fetchurl a thing. Even more, this wasn't explicitly proposed _here_, but I've implied that, now _source_ is an argument for an expression:

```
{ stdenv, fetchurl, python2Packages, makeWrapper, unzip, ...
,  mercurialSrc ? fetchurl rec {
    meta.name = "mercurial-${meta.version}";
    meta.version = "4.9.1";
    url = "https://mercurial-scm.org/release/${meta.name}.tar.gz";
    sha256 = "0iybbkd9add066729zg01kwz5hhc1s6lhp9rrnsmzq6ihyxj3p8v";
  }
}:
python2Packages.buildPythonApplication {

  inherit (mercurialSrc.meta) name version;
  src = mercurialSrc;
  ...
```
This change now allows overriding `src`, `name` and `version` through `.override {}`, without the need to call `.overrideAttrs`.
